### PR TITLE
Validate External Tracker URL Format

### LINKED ISSUE #7068: Check tracker url syntax
- Gitea version (or commit ref): 81adf6a
- Git version:
- Operating system:
- Database (use `[x]`):
  - [ ] PostgreSQL
  - [x] MySQL
  - [ ] MSSQL
  - [ ] SQLite
- Can you reproduce the bug at https://try.gitea.io:
  - [x] Yes (provide example URL)
  - [ ] No
  - [ ] Not relevant
- Log gist:

## Description
Some time ago we created a new repository and added a connection to our external issue tracking system Tuleap.
We used this repository a while and didn´t recognized that there was a problem with the tracking url.

Yesterday i pushed a commit containing a tracker id in the commit message for the first time. After this, it wasn´t possible to view the repository in the web view anymore.
Because of this i wasn´t able to review or reconfigure the repository settings. I could only delete the whole repository via admin settings.

Web output in repository view:
`template: repo/view_list:30:35: executing "repo/view_list" at <RenderCommitMessage .LatestCommit.Summary .RepoLink $.Repository.ComposeMetas>: error calling RenderCommitMessage: runtime error: slice bounds out of range`
and in diff view:
`template: repo/diff/page:12:28: executing "repo/diff/page" at <RenderCommitMessage .Commit.Message $.RepoLink $.Repository.ComposeMetas>: error calling RenderCommitMessage: runtime error: slice bounds out of range`

In the gitea db i found a table called repo_unit with configuration data of the repos.
In this case the entry was 
`{"ExternalTrackerURL":"https://sampledomain.com/plugins/tracker/?tracker=161","ExternalTrackerFormat":"https://sampledomain.com/plugins/tracker/?aid={index]","ExternalTrackerStyle":"numeric"}`
My colleague did a mistake on closing brackets > `?aid={index]`

I suggest the web interface should perform an error check on entering the external tracker url to prevent such problems.

On try.gitea.io the problem seems to only occur on diff view and graph so this problem can be fixed in the web interface if you know what the problem is.
see here https://try.gitea.io/Grelog/test1/commit/643c2b265183847ace82967542ce40ac2ce92a7f
and here https://try.gitea.io/Grelog/test1/graph

### PR DESCRIPTION
**Pull request by mrsdizzie**
Add some validation checks for external tracker URL format.

Attempts to check for simple typos by making sure there is a matching open/close brace for placeholder variable names if the opposite is present.

Fixes #7068

### CODE REVIEW

File: modules/validation/helpers.go [hunk 1]
```diff
@@ -75,3 +76,21 @@ func IsValidExternalURL(uri string) bool {
...
 	return true
 }
+
+// IsValidExternalTrackerURLFormat checks if URL matches required syntax for external trackers
+func IsValidExternalTrackerURLFormat(uri string) bool {
+	if !IsValidExternalURL(uri) {
+		return false
+	}
+
+	var re = regexp.MustCompile(`({?)(?:user|repo|index)+?(}?)`)
```
Comment (adelowo) [inline modules/validation/helpers.go]: Can we move this to a package level variable so it is compiled only once?

**Review by adelowo — CHANGES_REQUESTED** [review]

File: modules/validation/helpers.go [hunk 2]
```diff
@@ -1,4 +1,4 @@
-// Copyright 2018 The Gitea Authors. All rights reserved.
```
Comment (adelowo) [inline modules/validation/helpers.go]: I am quite certain it is fine not to update this year... Maybe @techknowlogick knows better?

File: modules/validation/helpers.go (hunk 1 above)
Comment (mrsdizzie) [reply modules/validation/helpers.go]: sure, moved!

File: modules/validation/helpers.go (hunk 2 above)
Comment (techknowlogick) [reply modules/validation/helpers.go]: We should probably think about our copyright header update policy, but as of right now the date is of when the first change to the file is made.

Comment (mrsdizzie) [reply modules/validation/helpers.go]: ah OK, reverted that thx

**Review by adelowo — APPROVED** [review]

**Review by lunny — APPROVED** [review]

**Review by techknowlogick — APPROVED** [review]

**Pull request merged**